cmd/snap-bootstrap/initramfs_install_handler.go

```diff
@@ -0,0 +1,418 @@
+		BasePath:          seedSnaps[snap.TypeBase].Path,
+		Gadget:            snapInfos[snap.TypeGadget],
+		GadgetPath:        seedSnaps[snap.TypeGadget].Path,
+		Kernel:            snapInfos[snap.TypeKernel],
+		KernelPath:        seedSnaps[snap.TypeKernel].Path,
+		UnpackedGadgetDir: gadgetMount,
+
+		RecoverySystemLabel: recoverySystemLabel,
+	}
+
+	// wire fde hook from initrd
+	boot.HasFDESetupHook = hasFDESetupHook
+	boot.RunFDESetupHook = runFDESetupHook
+	if err = boot.MakeRunnableSystem(model, bootWith, trustedInstallObserver); err != nil {
+		return fmt.Errorf("cannot make system runnable: %v\n", err)
+	}
+
+	logger.Noticef("Installing system took %s\n", time.Since(start))
+	return nil
+}
+
+func maybeApplyPreseededData(deviceSeed seed.Seed, assertDb *asserts.Database, systemLabel string) error {
+	start := time.Now()
+	preseedArtifact := filepath.Join(boot.InitramfsUbuntuSeedDir, "systems", systemLabel, "preseed.tgz")
+	if !osutil.FileExists(preseedArtifact) {
+		logger.Noticef("preseed data not present, will do normal seeding")
+		return nil
+	}
+	model := deviceSeed.Model()
+	tmpDb := assertDb.WithStackedBackstore(asserts.NewMemoryBackstore())
+	preseedAs, err := devicestate.InstallHandlerDoReadPreseedAssertion(tmpDb, model, boot.InitramfsUbuntuSeedDir, systemLabel)
+	if err != nil {
+		return fmt.Errorf("failed to read preseed assertion: %v", err)
+	}
+	writableDir := boot.InstallHostWritableDir(model)
+	if err != nil {
+		return err
+	}
+	// TODO: improve handling of the preseed, we do double hashing for already loaded snaps
+	// at the same time, we need to process them to generarate dir structures e.g. /sna/<name>/<rev>
+	if err = devicestate.InstallHandlerDoMaybeApplyPreseededData(deviceSeed.Model(),
+		preseedAs, preseedArtifact, boot.InitramfsUbuntuSeedDir, systemLabel, writableDir); err != nil {
+		return fmt.Errorf("failed to apply preseed data: %v", err)
+	}
+	logger.Noticef("successfully preseeded the system")
+	logger.Noticef("Preseeded data handled in %s\n", time.Since(start))
+	return nil
+}
+
+func prepareUbuntuRuntimeMounts(deviceSeed seed.Seed) error {
+	// unmount gadget and kernel (currently mounted from seed)
+	if err := umountEssentials(); err != nil {
+		return err
+	}
+
+	// TODO: ubuntu-seed/boot/save are already mounted and at right location
+	// move /run/mnt/ubuntu-data to /run/mnt/data
+	// we cannot move mount, instead we need to unmount -> remount
+	// if err := sd.Mount(InstallUbuntuDataDir, boot.InitramfsDataDir, "-o", "move"); err != nil {
+	// 	logger.Noticef("Failed to move ubuntu-data mount..%v", err)
+	// 	return err
+	// }
+	sd := systemd.New(systemd.SystemMode, progress.Null)
+
+	mountInfos, err := osutil.LoadMountInfo()
+	if err != nil {
+		return fmt.Errorf("failed to load mount info: %v", err)
+	}
+	ubuntuDataDevice := ""
+	for _, mnt := range mountInfos {
+		if mnt.MountDir == InstallUbuntuDataDir {
+			ubuntuDataDevice = mnt.MountSource
+			break
+		}
+	}
```

Comment:
this looks a bit odd, there should be a more direct to know this

Reply:
I'd love a simpler way here, but I could not find a go helper to get single-mount info.

Ultimately I do not think we will need this.
We divided `/run/mnt/data` and `/run/mnt/ubuntu-data` because we boot into install mode, where `/run/mnt/data` is `tmpfs` but this is no more concern. So my preference would be to run the install directly to `/run/mnt/data` and this extra dance is no more needed.